pilot/pkg/serviceregistry/kube/controller/pod.go

```diff
@@ -148,8 +148,18 @@ func (pc *PodCache) onEvent(old, pod *v1.Pod, ev model.Event) error {
 	ip := pod.Status.PodIP
 	// PodIP will be empty when pod is just created, but before the IP is assigned
 	// via UpdateStatus.
+	// In the case of a pod being created, we should not add it to the cache until it is ready.
+	// However, if the pod *used to* have an IP, we do need to actually delete it.
 	if len(ip) == 0 {
```

Comment:
IIUC, the ip is missing when being evicted. Then you can allow ip unset if this is a deletion event

Reply:
But the problem is that all the caches (podCache, and endpointShardz) are indexed by podIP, not pod name.

Reply:
Isn't there a map ipByPods map[types.NamespacedName]string

Reply:
Whevere this effect is same, but checking delete may reduce map quering

Reply:
Sorry, missed this! But yes, this PR checks in ipByPods, but even for "non deletions" (i.e. Failed), we need to check in the map. So we only do so if len(ip) == 0 already.